File: java/src/main/java/vsdk/toolkit/galerkin/processing/LinkingSimpleStrategy.java
package vsdk.toolkit.galerkin.processing;

import java.util.ArrayList;
import vsdk.toolkit.common.logging.Logger;
import vsdk.toolkit.galerkin.GalerkinBasis;
import vsdk.toolkit.galerkin.GalerkinElement;
import vsdk.toolkit.galerkin.GalerkinIterationMethod;
import vsdk.toolkit.galerkin.GalerkinRole;
import vsdk.toolkit.galerkin.GalerkinShaftCullMode;
import vsdk.toolkit.galerkin.GalerkinState;
import vsdk.toolkit.galerkin.Interaction;
import vsdk.toolkit.galerkin.Shaft;
import vsdk.toolkit.scene.Polygon;
import vsdk.toolkit.scene.Scene;
import vsdk.toolkit.skin.BoundingBox;
import vsdk.toolkit.skin.ElementFlags;
import vsdk.toolkit.skin.Geometry;
import vsdk.toolkit.skin.Patch;

public class LinkingSimpleStrategy {
    private static final ArrayList<float[]> coefficientsPool = new ArrayList<>();

    private static float[] borrowKBuffer() {
        int n = coefficientsPool.size();
        if ( n > 0 ) {
            return coefficientsPool.remove(n - 1);
        }
        return new float[GalerkinBasis.MAX_BASIS_SIZE * GalerkinBasis.MAX_BASIS_SIZE];
    }

    private static void returnKBuffer(float[] buffer) {
        if ( buffer != null && buffer.length == GalerkinBasis.MAX_BASIS_SIZE * GalerkinBasis.MAX_BASIS_SIZE ) {
            coefficientsPool.add(buffer);
        }
    }

    private static void createInitialLink(
        Scene scene,
        GalerkinState galerkinState,
        GalerkinRole role,
        ArrayList<Geometry>[] candidateList,
        GalerkinElement topElement,
        BoundingBox topLevelBoundingBox,
        Patch patch)
    {
        if ( scene == null || galerkinState == null || candidateList == null || candidateList.length == 0
             || topElement == null || patch == null || topElement.patch == null ) {
            return;
        }

        if ( !patch.facing(topElement.patch) ) {
            return;
        }

        GalerkinElement receiverElement;
        GalerkinElement sourceElement;
        GalerkinElement topLevelElement = GalerkinElement.fromPatch(patch);

        switch ( role ) {
            case SOURCE:
                receiverElement = topLevelElement;
                sourceElement = topElement;
                break;
            case RECEIVER:
                receiverElement = topElement;
                sourceElement = topLevelElement;
                break;
            default:
                Logger.fatal(2, "createInitialLink", "Impossible element role");
                return;
        }

        ArrayList<Geometry> oldCandidateList = candidateList[0];
        if ( (galerkinState.exactVisibility != 0
              || galerkinState.shaftCullMode == GalerkinShaftCullMode.ALWAYS_DO_SHAFT_CULLING)
             && oldCandidateList != null ) {
            Shaft shaft = new Shaft();

            if ( galerkinState.exactVisibility != 0 ) {
                if ( receiverElement != null && sourceElement != null ) {
                    Polygon receiverPolygon = new Polygon();
                    Polygon sourcePolygon = new Polygon();
                    receiverElement.initPolygon(receiverPolygon);
                    sourceElement.initPolygon(sourcePolygon);
                    shaft.constructFromPolygonToPolygon(receiverPolygon, sourcePolygon);
                }
            }
            else {
                BoundingBox boundingBox = new BoundingBox();
                patch.computeAndGetBoundingBox(boundingBox);
                shaft.constructFromBoundingBoxes(topLevelBoundingBox, boundingBox);
            }

            shaft.setShaftOmit(topElement.patch);
            shaft.setShaftOmit(patch);
            ArrayList<Geometry> arr = new ArrayList<>();
            shaft.doCulling(oldCandidateList, arr, galerkinState.shaftCullStrategy);
            candidateList[0] = arr;

            if ( shaft.isCut() ) {
                Shaft.freeCandidateList(candidateList[0]);
                candidateList[0] = oldCandidateList;
                return;
            }
        }

        Interaction link = new Interaction();
        link.K = borrowKBuffer();
        link.ownsK = false;
        link.receiverElement = receiverElement;
        link.sourceElement = sourceElement;

        if ( receiverElement != null ) {
            link.numberOfBasisFunctionsOnReceiver = receiverElement.basisSize;
        }

        if ( sourceElement != null ) {
            link.numberOfBasisFunctionsOnSource = sourceElement.basisSize;
        }

        boolean isSceneGeometry = candidateList[0] == scene.geometryList;
        boolean isClusteredGeometry = candidateList[0] == scene.clusteredGeometryList;
        ArrayList<Geometry> geometryListReferences = candidateList[0];
        FormFactorStrategy.computeAreaToAreaFormFactorVisibility(
            scene.voxelGrid,
            geometryListReferences,
            isSceneGeometry,
            isClusteredGeometry,
            link,
            galerkinState);

        if ( galerkinState.exactVisibility != 0
             || galerkinState.shaftCullMode == GalerkinShaftCullMode.ALWAYS_DO_SHAFT_CULLING ) {
            if ( oldCandidateList != candidateList[0] ) {
                Shaft.freeCandidateList(candidateList[0]);
            }
            candidateList[0] = oldCandidateList;
        }

        if ( (link.visibility & 0xFF) > 0 ) {
            Interaction newLink = Interaction.interactionDuplicate(link);
            if ( newLink == null ) {
                return;
            }

            // Store interactions with the source patch for the progressive radiosity
            // method and with the receiving patch for gathering methods.
            if ( galerkinState.galerkinIterationMethod == GalerkinIterationMethod.SOUTH_WELL ) {
                if ( sourceElement != null ) {
                    if ( sourceElement.interactions == null ) {
                        sourceElement.interactions = new ArrayList<>();
                    }
                    sourceElement.interactions.add(newLink);
                }
            }
            else if ( receiverElement != null ) {
                if ( receiverElement.interactions == null ) {
                    receiverElement.interactions = new ArrayList<>();
                }
                receiverElement.interactions.add(newLink);
            }
        }

        float[] borrowed = link.K;
        link.K = null;
        returnKBuffer(borrowed);
    }

    private static void geometryLink(
        Scene scene,
        GalerkinState galerkinState,
        GalerkinRole role,
        ArrayList<Geometry>[] candidateList,
        GalerkinElement topElement,
        BoundingBox topLevelBoundingBox,
        Geometry geometry)
    {
        if ( scene == null || galerkinState == null || candidateList == null || candidateList.length == 0
             || topElement == null || topElement.patch == null || geometry == null ) {
            return;
        }

        // Immediately return if the geometry is bounded and behind the plane
        // of the patch for which interactions are created.
        if ( geometry.bounded
             && geometry.getBoundingBox().behindPlane(topElement.patch.normal, topElement.patch.planeConstant) ) {
            return;
        }

        // If the geometry is bounded, do shaft culling, reducing the candidate
        // list which contains possible occluders between patch pairs.
        ArrayList<Geometry> oldCandidateList = candidateList[0];
        if ( geometry.bounded && oldCandidateList != null ) {
            Shaft shaft = new Shaft();
            BoundingBox boundingBox = geometry.getBoundingBox();
            shaft.constructFromBoundingBoxes(topLevelBoundingBox, boundingBox);
            shaft.setShaftOmit(topElement.patch);
            ArrayList<Geometry> arr = new ArrayList<>();
            shaft.doCulling(oldCandidateList, arr, galerkinState.shaftCullStrategy);
            candidateList[0] = arr;
        }

        // If the geometry is an aggregate, test each child. If it is primitive,
        // create an initial link with each patch it consists of.
        if ( geometry.isCompound() ) {
            ArrayList<Geometry> geometryList = Geometry.primitiveListCopy(geometry);
            for ( int i = 0; geometryList != null && i < geometryList.size(); i++ ) {
                geometryLink(
                    scene,
                    galerkinState,
                    role,
                    candidateList,
                    topElement,
                    topLevelBoundingBox,
                    geometryList.get(i));
            }
        }
        else {
            ArrayList<Patch> patchList = Geometry.patchListReference(geometry);
            for ( int i = 0; patchList != null && i < patchList.size(); i++ ) {
                createInitialLink(
                    scene,
                    galerkinState,
                    role,
                    candidateList,
                    topElement,
                    topLevelBoundingBox,
                    patchList.get(i));
            }
        }

        if ( geometry.bounded && oldCandidateList != null ) {
            Shaft.freeCandidateList(candidateList[0]);
        }
        candidateList[0] = oldCandidateList;
    }

    public static void createInitialLinks(
        Scene scene,
        GalerkinState galerkinState,
        GalerkinRole role,
        GalerkinElement topElement)
    {
        if ( scene == null || galerkinState == null || topElement == null ) {
            return;
        }

        if ( (topElement.flags & ElementFlags.IS_CLUSTER_MASK) != 0 ) {
            Logger.fatal(-1, "createInitialLinks", "cannot use this routine for cluster elements");
            return;
        }
        if ( topElement.patch == null ) {
            return;
        }

        BoundingBox topLevelBoundingBox = new BoundingBox();
        topElement.patch.computeAndGetBoundingBox(topLevelBoundingBox);

        @SuppressWarnings("unchecked")
        ArrayList<Geometry>[] candidateList = (ArrayList<Geometry>[])new ArrayList<?>[1];
        candidateList[0] = scene.clusteredGeometryList;

        for ( int i = 0; scene.geometryList != null && i < scene.geometryList.size(); i++ ) {
            geometryLink(
                scene,
                galerkinState,
                role,
                candidateList,
                topElement,
                topLevelBoundingBox,
                scene.geometryList.get(i));
        }
    }
}

File: typescript/src/main/vsdk/toolkit/galerkin/processing/LinkingSimpleStrategy.ts
import { Error as VsdkError } from "../../common/Error";
import { GalerkinBasis } from "../GalerkinBasis";
import { GalerkinElement } from "../GalerkinElement";
import { GalerkinIterationMethod } from "../GalerkinIterationMethod";
import { GalerkinRole } from "../GalerkinRole";
import { GalerkinShaftCullMode } from "../GalerkinShaftCullMode";
import { GalerkinState } from "../GalerkinState";
import { Interaction } from "../Interaction";
import { Shaft } from "../Shaft";
import { Polygon } from "../../scene/Polygon";
import { Scene } from "../../scene/Scene";
import { BoundingBox } from "../../skin/BoundingBox";
import { ElementFlags } from "../../skin/ElementFlags";
import { Geometry } from "../../skin/Geometry";
import { Patch } from "../../skin/Patch";
import { FormFactorStrategy } from "./FormFactorStrategy";

export class LinkingSimpleStrategy {
  private static createInitialLink(
    scene: Scene,
    galerkinState: GalerkinState,
    role: GalerkinRole,
    candidateList: Array<Geometry[] | null>,
    topElement: GalerkinElement,
    topLevelBoundingBox: BoundingBox,
    patch: Patch,
  ): void {
    if (scene === null || galerkinState === null || candidateList === null || candidateList.length === 0
      || topElement === null || patch === null || topElement.patch === null) {
      return;
    }

    if (!patch.facing(topElement.patch)) {
      return;
    }

    let receiverElement: GalerkinElement;
    let sourceElement: GalerkinElement;
    const topLevelElement = GalerkinElement.fromPatch(patch);
    if (topLevelElement === null) {
      return;
    }

    switch (role) {
      case GalerkinRole.SOURCE:
        receiverElement = topLevelElement;
        sourceElement = topElement;
        break;
      case GalerkinRole.RECEIVER:
        receiverElement = topElement;
        sourceElement = topLevelElement;
        break;
      default:
        VsdkError.fatal(2, "createInitialLink", "Impossible element role");
        return;
    }

    const oldCandidateList = candidateList[0];
    if ((galerkinState.exactVisibility !== 0
        || galerkinState.shaftCullMode === GalerkinShaftCullMode.ALWAYS_DO_SHAFT_CULLING)
      && oldCandidateList !== null) {
      const shaft = new Shaft();

      if (galerkinState.exactVisibility !== 0) {
        const receiverPolygon = new Polygon();
        const sourcePolygon = new Polygon();
        receiverElement.initPolygon(receiverPolygon);
        sourceElement.initPolygon(sourcePolygon);
        shaft.constructFromPolygonToPolygon(receiverPolygon, sourcePolygon);
      }
      else {
        const boundingBox = new BoundingBox();
        patch.computeAndGetBoundingBox(boundingBox);
        shaft.constructFromBoundingBoxes(topLevelBoundingBox, boundingBox);
      }

      shaft.setShaftOmit(topElement.patch);
      shaft.setShaftOmit(patch);
      const arr: Geometry[] = [];
      shaft.doCulling(oldCandidateList, arr, galerkinState.shaftCullStrategy);
      candidateList[0] = arr;

      if (shaft.isCut()) {
        Shaft.freeCandidateList(candidateList[0]);
        candidateList[0] = oldCandidateList;
        return;
      }
    }

    const link = new Interaction();
    link.K = new Array<number>(GalerkinBasis.MAX_BASIS_SIZE * GalerkinBasis.MAX_BASIS_SIZE).fill(0.0);
    link.receiverElement = receiverElement;
    link.sourceElement = sourceElement;

    link.numberOfBasisFunctionsOnReceiver = receiverElement.basisSize;
    link.numberOfBasisFunctionsOnSource = sourceElement.basisSize;

    const isSceneGeometry = candidateList[0] === scene.geometryList;
    const isClusteredGeometry = candidateList[0] === scene.clusteredGeometryList;
    const geometryListReferences = candidateList[0];
    FormFactorStrategy.computeAreaToAreaFormFactorVisibility(
      scene.voxelGrid,
      geometryListReferences,
      isSceneGeometry,
      isClusteredGeometry,
      link,
      galerkinState,
    );

    if (galerkinState.exactVisibility !== 0
      || galerkinState.shaftCullMode === GalerkinShaftCullMode.ALWAYS_DO_SHAFT_CULLING) {
      if (oldCandidateList !== candidateList[0]) {
        Shaft.freeCandidateList(candidateList[0]);
      }
      candidateList[0] = oldCandidateList;
    }

    if ((link.visibility & 0xff) > 0) {
      const newLink = Interaction.interactionDuplicate(link);
      if (newLink === null) {
        return;
      }

      if (galerkinState.galerkinIterationMethod === GalerkinIterationMethod.SOUTH_WELL) {
        if (sourceElement.interactions === null) {
          sourceElement.interactions = [];
        }
        sourceElement.interactions.push(newLink);
      }
      else {
        if (receiverElement.interactions === null) {
          receiverElement.interactions = [];
        }
        receiverElement.interactions.push(newLink);
      }
    }
  }

  private static geometryLink(
    scene: Scene,
    galerkinState: GalerkinState,
    role: GalerkinRole,
    candidateList: Array<Geometry[] | null>,
    topElement: GalerkinElement,
    topLevelBoundingBox: BoundingBox,
    geometry: Geometry,
  ): void {
    if (scene === null || galerkinState === null || candidateList === null || candidateList.length === 0
      || topElement === null || topElement.patch === null || geometry === null) {
      return;
    }

    if (geometry.bounded
      && geometry.getBoundingBox().behindPlane(topElement.patch.normal, topElement.patch.planeConstant)) {
      return;
    }

    const oldCandidateList = candidateList[0];
    if (geometry.bounded && oldCandidateList !== null) {
      const shaft = new Shaft();
      const boundingBox = geometry.getBoundingBox();
      shaft.constructFromBoundingBoxes(topLevelBoundingBox, boundingBox);
      shaft.setShaftOmit(topElement.patch);
      const arr: Geometry[] = [];
      shaft.doCulling(oldCandidateList, arr, galerkinState.shaftCullStrategy);
      candidateList[0] = arr;
    }

    if (geometry.isCompound()) {
      const geometryList = Geometry.primitiveListCopy(geometry);
      for (let i = 0; geometryList !== null && i < geometryList.length; i++) {
        LinkingSimpleStrategy.geometryLink(
          scene,
          galerkinState,
          role,
          candidateList,
          topElement,
          topLevelBoundingBox,
          geometryList[i],
        );
      }
    }
    else {
      const patchList = Geometry.patchListReference(geometry);
      for (let i = 0; patchList !== null && i < patchList.length; i++) {
        LinkingSimpleStrategy.createInitialLink(
          scene,
          galerkinState,
          role,
          candidateList,
          topElement,
          topLevelBoundingBox,
          patchList[i],
        );
      }
    }

    if (geometry.bounded && oldCandidateList !== null) {
      Shaft.freeCandidateList(candidateList[0]);
    }
    candidateList[0] = oldCandidateList;
  }

  public static createInitialLinks(
    scene: Scene,
    galerkinState: GalerkinState,
    role: GalerkinRole,
    topElement: GalerkinElement,
  ): void {
    if (scene === null || galerkinState === null || topElement === null) {
      return;
    }

    if ((topElement.flags & ElementFlags.IS_CLUSTER_MASK) !== 0) {
      VsdkError.fatal(-1, "createInitialLinks", "cannot use this routine for cluster elements");
      return;
    }
    if (topElement.patch === null) {
      return;
    }

    const topLevelBoundingBox = new BoundingBox();
    topElement.patch.computeAndGetBoundingBox(topLevelBoundingBox);

    const candidateList: Array<Geometry[] | null> = [scene.clusteredGeometryList];

    for (let i = 0; scene.geometryList !== null && i < scene.geometryList.length; i++) {
      LinkingSimpleStrategy.geometryLink(
        scene,
        galerkinState,
        role,
        candidateList,
        topElement,
        topLevelBoundingBox,
        scene.geometryList[i],
      );
    }
  }
}
